<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Case Studies in Reproducible Research: a spring seminar at UCSC</title>
  <meta name="description" content="This is a bookdown that stores materials for a UCSC seminar on reproducible research with R. The output format for this example is bookdown::gitbook.">
  <meta name="generator" content="bookdown 0.3.14 and GitBook 2.6.7">

  <meta property="og:title" content="Case Studies in Reproducible Research: a spring seminar at UCSC" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a bookdown that stores materials for a UCSC seminar on reproducible research with R. The output format for this example is bookdown::gitbook." />
  <meta name="github-repo" content="eriqande/rep-res-eeb-2017" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Case Studies in Reproducible Research: a spring seminar at UCSC" />
  
  <meta name="twitter:description" content="This is a bookdown that stores materials for a UCSC seminar on reproducible research with R. The output format for this example is bookdown::gitbook." />
  

<meta name="author" content="Eric C. Anderson, Kristen C. Ruegg, Tina Cheng, and the students of EEB 295">


<meta name="date" content="2017-06-02">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="week4.html">
<link rel="next" href="ggplot2-intro-lecture.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Case Studies in Reproducible Research</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Course Overview</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#meeting-times-location-requirements"><i class="fa fa-check"></i><b>1.1</b> Meeting Times, Location, Requirements</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#the-origin-of-this-seminar"><i class="fa fa-check"></i><b>1.2</b> The origin of this seminar</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#course-organizers"><i class="fa fa-check"></i><b>1.3</b> Course Organizers</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#course-goals"><i class="fa fa-check"></i><b>1.4</b> Course Goals</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#weekly-syllabus"><i class="fa fa-check"></i><b>1.5</b> Weekly Syllabus</a><ul>
<li class="chapter" data-level="1.5.1" data-path="index.html"><a href="index.html#week-1-introduction-and-getting-your-workspace-set-up"><i class="fa fa-check"></i><b>1.5.1</b> Week 1 — Introduction and Getting Your Workspace Set Up</a></li>
<li class="chapter" data-level="1.5.2" data-path="index.html"><a href="index.html#week-2-rstudio-project-organization-using-git-and-github-quick-rmarkdown"><i class="fa fa-check"></i><b>1.5.2</b> Week 2 — RStudio project organization; using git and GitHub; Quick RMarkdown</a></li>
<li class="chapter" data-level="1.5.3" data-path="index.html"><a href="index.html#week-3-tibbles.-reading-data-in.-data-rectangling"><i class="fa fa-check"></i><b>1.5.3</b> Week 3 — Tibbles. Reading data in. Data rectangling</a></li>
<li class="chapter" data-level="1.5.4" data-path="index.html"><a href="index.html#week-4"><i class="fa fa-check"></i><b>1.5.4</b> Week 4 —</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="week1.html"><a href="week1.html"><i class="fa fa-check"></i><b>2</b> Week One Meeting</a><ul>
<li class="chapter" data-level="2.1" data-path="week1.html"><a href="week1.html#software-installation"><i class="fa fa-check"></i><b>2.1</b> Software Installation</a></li>
<li class="chapter" data-level="2.2" data-path="week1.html"><a href="week1.html#get-an-account-on-github"><i class="fa fa-check"></i><b>2.2</b> Get an account on GitHub</a><ul>
<li class="chapter" data-level="2.2.1" data-path="week1.html"><a href="week1.html#private-repositories"><i class="fa fa-check"></i><b>2.2.1</b> Private repositories</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="week1.html"><a href="week1.html#open-an-rstudio-project-from-github"><i class="fa fa-check"></i><b>2.3</b> Open an RStudio Project from GitHub</a></li>
<li class="chapter" data-level="2.4" data-path="week1.html"><a href="week1.html#assignment-for-next-week-create-an-rstudio-project-with-your-own-data"><i class="fa fa-check"></i><b>2.4</b> Assignment for next week: Create an RStudio Project with Your Own Data</a></li>
<li class="chapter" data-level="2.5" data-path="week1.html"><a href="week1.html#week2reading"><i class="fa fa-check"></i><b>2.5</b> Reading for next week</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="week2.html"><a href="week2.html"><i class="fa fa-check"></i><b>3</b> Week Two Meeting</a><ul>
<li class="chapter" data-level="3.1" data-path="week2.html"><a href="week2.html#workflow-and-project-recap"><i class="fa fa-check"></i><b>3.1</b> Workflow and Project Recap</a><ul>
<li class="chapter" data-level="3.1.1" data-path="week2.html"><a href="week2.html#workflow-basics"><i class="fa fa-check"></i><b>3.1.1</b> Workflow: basics</a></li>
<li class="chapter" data-level="3.1.2" data-path="week2.html"><a href="week2.html#workflow-projects"><i class="fa fa-check"></i><b>3.1.2</b> Workflow: projects</a></li>
<li class="chapter" data-level="3.1.3" data-path="week2.html"><a href="week2.html#workflow-scripts"><i class="fa fa-check"></i><b>3.1.3</b> Workflow: scripts</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="week2.html"><a href="week2.html#lets-talk-about-the-pipe"><i class="fa fa-check"></i><b>3.2</b> Let’s talk about the pipe <code>%&gt;%</code></a></li>
<li class="chapter" data-level="3.3" data-path="week2.html"><a href="week2.html#tibbles-and-rectangular-data"><i class="fa fa-check"></i><b>3.3</b> Tibbles and “rectangular” data</a><ul>
<li class="chapter" data-level="3.3.1" data-path="week2.html"><a href="week2.html#tibble-excercises"><i class="fa fa-check"></i><b>3.3.1</b> Tibble excercises</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="week2.html"><a href="week2.html#data-import"><i class="fa fa-check"></i><b>3.4</b> Data import</a><ul>
<li class="chapter" data-level="3.4.1" data-path="week2.html"><a href="week2.html#rstudios-gui-importer"><i class="fa fa-check"></i><b>3.4.1</b> RStudio’s GUI importer</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="week3.html"><a href="week3.html"><i class="fa fa-check"></i><b>4</b> Week Three Meeting</a><ul>
<li class="chapter" data-level="4.1" data-path="week3.html"><a href="week3.html#git-basics"><i class="fa fa-check"></i><b>4.1</b> Git Basics</a><ul>
<li class="chapter" data-level="4.1.1" data-path="week3.html"><a href="week3.html#vcs"><i class="fa fa-check"></i><b>4.1.1</b> An overview of Version Control Systems (VCS)</a></li>
<li class="chapter" data-level="4.1.2" data-path="week3.html"><a href="week3.html#git-versus-github"><i class="fa fa-check"></i><b>4.1.2</b> Git versus GitHub</a></li>
<li class="chapter" data-level="4.1.3" data-path="week3.html"><a href="week3.html#git-thru-rstudio"><i class="fa fa-check"></i><b>4.1.3</b> Using git through RStudio</a></li>
<li class="chapter" data-level="4.1.4" data-path="week3.html"><a href="week3.html#git-play"><i class="fa fa-check"></i><b>4.1.4</b> Go for it everyone!</a></li>
<li class="chapter" data-level="4.1.5" data-path="week3.html"><a href="week3.html#git-how"><i class="fa fa-check"></i><b>4.1.5</b> How does git store and keep track of things</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="week3.html"><a href="week3.html#pushing-and-pulling-with-github"><i class="fa fa-check"></i><b>4.2</b> Pushing and Pulling With GitHub</a><ul>
<li class="chapter" data-level="4.2.1" data-path="week3.html"><a href="week3.html#creating-a-repository-on-github-and-the-initial-push"><i class="fa fa-check"></i><b>4.2.1</b> Creating a Repository on GitHub and the initial push</a></li>
<li class="chapter" data-level="4.2.2" data-path="week3.html"><a href="week3.html#subsequent-pushes"><i class="fa fa-check"></i><b>4.2.2</b> Subsequent pushes</a></li>
<li class="chapter" data-level="4.2.3" data-path="week3.html"><a href="week3.html#assign-collaborators"><i class="fa fa-check"></i><b>4.2.3</b> Assign collaborators</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="week3.html"><a href="week3.html#next-weeks-assignment"><i class="fa fa-check"></i><b>4.3</b> Next Week’s Assignment</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="week4.html"><a href="week4.html"><i class="fa fa-check"></i><b>5</b> Week Four Meeting</a><ul>
<li class="chapter" data-level="5.1" data-path="week4.html"><a href="week4.html#knit-your-readme.rmd-files"><i class="fa fa-check"></i><b>5.1</b> Knit your README.Rmd files</a></li>
<li class="chapter" data-level="5.2" data-path="week4.html"><a href="week4.html#changing-to-factors"><i class="fa fa-check"></i><b>5.2</b> Changing to factors</a></li>
<li class="chapter" data-level="5.3" data-path="week4.html"><a href="week4.html#the-group_by-function"><i class="fa fa-check"></i><b>5.3</b> The <code>group_by()</code> function</a></li>
<li class="chapter" data-level="5.4" data-path="week4.html"><a href="week4.html#how-do-i-learn-about-all-the-vectorized-functions-i-can-use-in-mutate-and-summarize"><i class="fa fa-check"></i><b>5.4</b> How do I learn about all the vectorized functions I can use in <code>mutate()</code> and <code>summarize()</code>?</a></li>
<li class="chapter" data-level="5.5" data-path="week4.html"><a href="week4.html#next-weeks-assignment-1"><i class="fa fa-check"></i><b>5.5</b> Next Week’s Assignment</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="week5.html"><a href="week5.html"><i class="fa fa-check"></i><b>6</b> Week Five Meeting</a><ul>
<li class="chapter" data-level="6.1" data-path="week5.html"><a href="week5.html#some-things-regarding-peoples-repositories"><i class="fa fa-check"></i><b>6.1</b> Some things regarding people’s repositories</a><ul>
<li class="chapter" data-level="6.1.1" data-path="week5.html"><a href="week5.html#data-compression"><i class="fa fa-check"></i><b>6.1.1</b> Data Compression</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="week5.html"><a href="week5.html#a-quick-aside-about-missing-data"><i class="fa fa-check"></i><b>6.2</b> A quick aside about missing data</a></li>
<li class="chapter" data-level="6.3" data-path="week5.html"><a href="week5.html#brief-highlights-of-the-joins-chapter"><i class="fa fa-check"></i><b>6.3</b> Brief Highlights of the Joins Chapter</a><ul>
<li class="chapter" data-level="6.3.1" data-path="week5.html"><a href="week5.html#a-few-thoughts-on-keys"><i class="fa fa-check"></i><b>6.3.1</b> A few thoughts on keys</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="week5.html"><a href="week5.html#an-example-of-using-some-joins"><i class="fa fa-check"></i><b>6.4</b> An example of using some joins</a><ul>
<li class="chapter" data-level="6.4.1" data-path="week5.html"><a href="week5.html#when-would-i-use-right_join"><i class="fa fa-check"></i><b>6.4.1</b> When would I use <code>right_join()</code>?</a></li>
<li class="chapter" data-level="6.4.2" data-path="week5.html"><a href="week5.html#what-about-that-inner_join"><i class="fa fa-check"></i><b>6.4.2</b> What about that <code>inner_join()</code></a></li>
<li class="chapter" data-level="6.4.3" data-path="week5.html"><a href="week5.html#an-anti_join-example"><i class="fa fa-check"></i><b>6.4.3</b> An anti_join example</a></li>
<li class="chapter" data-level="6.4.4" data-path="week5.html"><a href="week5.html#just-for-fun-lets-see-a-full_join"><i class="fa fa-check"></i><b>6.4.4</b> Just for fun, let’s see a full_join()</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="week5.html"><a href="week5.html#working-with-r-notebooks"><i class="fa fa-check"></i><b>6.5</b> Working with R Notebooks</a><ul>
<li class="chapter" data-level="6.5.1" data-path="week5.html"><a href="week5.html#open-your-own-r-notebook"><i class="fa fa-check"></i><b>6.5.1</b> Open your own R Notebook</a></li>
<li class="chapter" data-level="6.5.2" data-path="week5.html"><a href="week5.html#working-with-r-notebooks-1"><i class="fa fa-check"></i><b>6.5.2</b> Working with R Notebooks</a></li>
<li class="chapter" data-level="6.5.3" data-path="week5.html"><a href="week5.html#some-caveats-about-notebooks"><i class="fa fa-check"></i><b>6.5.3</b> Some caveats about notebooks</a></li>
<li class="chapter" data-level="6.5.4" data-path="week5.html"><a href="week5.html#opening-a-.nb.html-file-in-rstudio"><i class="fa fa-check"></i><b>6.5.4</b> Opening a .nb.html file in Rstudio</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="week5.html"><a href="week5.html#for-next-week"><i class="fa fa-check"></i><b>6.6</b> For Next Week:</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html"><i class="fa fa-check"></i><b>7</b> Week 6 meeting: using ggplot2</a><ul>
<li class="chapter" data-level="7.0.1" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#your-mission"><i class="fa fa-check"></i><b>7.0.1</b> Your Mission</a></li>
<li class="chapter" data-level="7.1" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#some-lecture-notes-from-a-few-years-ago"><i class="fa fa-check"></i><b>7.1</b> Some lecture notes from a few years ago</a><ul>
<li class="chapter" data-level="7.1.1" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#goals-for-this-hour"><i class="fa fa-check"></i><b>7.1.1</b> Goals for this hour:</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#about-ggplot2"><i class="fa fa-check"></i><b>7.2</b> About ggplot2</a><ul>
<li class="chapter" data-level="7.2.1" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#basics"><i class="fa fa-check"></i><b>7.2.1</b> Basics</a></li>
<li class="chapter" data-level="7.2.2" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#what-is-this-grammar-of-graphics"><i class="fa fa-check"></i><b>7.2.2</b> What is this grammar of graphics?</a></li>
<li class="chapter" data-level="7.2.3" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#ggplot2"><i class="fa fa-check"></i><b>7.2.3</b> ggplot2</a></li>
<li class="chapter" data-level="7.2.4" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#components-of-the-grammar-of-graphics"><i class="fa fa-check"></i><b>7.2.4</b> Components of the grammar of graphics</a></li>
<li class="chapter" data-level="7.2.5" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#in-a-nutshell"><i class="fa fa-check"></i><b>7.2.5</b> In a nutshell</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#pv-example"><i class="fa fa-check"></i><b>7.3</b> An example, please</a><ul>
<li class="chapter" data-level="7.3.1" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#a-pole-vaulting-example"><i class="fa fa-check"></i><b>7.3.1</b> A pole vaulting example</a></li>
<li class="chapter" data-level="7.3.2" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#a-first-ggplot"><i class="fa fa-check"></i><b>7.3.2</b> A first ggplot</a></li>
<li class="chapter" data-level="7.3.3" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#ggplots-system-of-defaults"><i class="fa fa-check"></i><b>7.3.3</b> ggplot’s system of defaults</a></li>
<li class="chapter" data-level="7.3.4" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#how-many-geoms-are-there"><i class="fa fa-check"></i><b>7.3.4</b> How many geoms are there?</a></li>
<li class="chapter" data-level="7.3.5" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#getting-even-sillier"><i class="fa fa-check"></i><b>7.3.5</b> Getting even sillier</a></li>
<li class="chapter" data-level="7.3.6" data-path="ggplot2-intro-lecture.html"><a href="ggplot2-intro-lecture.html#more-add-some-text-to-it"><i class="fa fa-check"></i><b>7.3.6</b> More! Add some text to it!</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="map-making-in-R.html"><a href="map-making-in-R.html"><i class="fa fa-check"></i><b>8</b> Week 7: Making Simple Maps with R</a><ul>
<li class="chapter" data-level="8.1" data-path="map-making-in-R.html"><a href="map-making-in-R.html#map-making-intro"><i class="fa fa-check"></i><b>8.1</b> Intro</a><ul>
<li class="chapter" data-level="8.1.1" data-path="map-making-in-R.html"><a href="map-making-in-R.html#todays-goals"><i class="fa fa-check"></i><b>8.1.1</b> Today’s Goals</a></li>
<li class="chapter" data-level="8.1.2" data-path="map-making-in-R.html"><a href="map-making-in-R.html#prerequisites"><i class="fa fa-check"></i><b>8.1.2</b> Prerequisites</a></li>
<li class="chapter" data-level="8.1.3" data-path="map-making-in-R.html"><a href="map-making-in-R.html#load-up-a-few-of-the-libraries-we-will-use"><i class="fa fa-check"></i><b>8.1.3</b> Load up a few of the libraries we will use</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="map-making-in-R.html"><a href="map-making-in-R.html#maps-package-and-ggplot"><i class="fa fa-check"></i><b>8.2</b> Plotting maps-package maps with ggplot</a><ul>
<li class="chapter" data-level="8.2.1" data-path="map-making-in-R.html"><a href="map-making-in-R.html#the-main-players"><i class="fa fa-check"></i><b>8.2.1</b> The main players:</a></li>
<li class="chapter" data-level="8.2.2" data-path="map-making-in-R.html"><a href="map-making-in-R.html#maps-in-the-maps-package"><i class="fa fa-check"></i><b>8.2.2</b> Maps in the maps package</a></li>
<li class="chapter" data-level="8.2.3" data-path="map-making-in-R.html"><a href="map-making-in-R.html#makin-data-frames-from-map-outlines"><i class="fa fa-check"></i><b>8.2.3</b> Makin’ data frames from map outlines</a></li>
<li class="chapter" data-level="8.2.4" data-path="map-making-in-R.html"><a href="map-making-in-R.html#the-structure-of-those-data-frames"><i class="fa fa-check"></i><b>8.2.4</b> The structure of those data frames</a></li>
<li class="chapter" data-level="8.2.5" data-path="map-making-in-R.html"><a href="map-making-in-R.html#plot-the-usa-map"><i class="fa fa-check"></i><b>8.2.5</b> Plot the USA map</a></li>
<li class="chapter" data-level="8.2.6" data-path="map-making-in-R.html"><a href="map-making-in-R.html#state-maps"><i class="fa fa-check"></i><b>8.2.6</b> State maps</a></li>
<li class="chapter" data-level="8.2.7" data-path="map-making-in-R.html"><a href="map-making-in-R.html#zoom-in"><i class="fa fa-check"></i><b>8.2.7</b> zoom in?</a></li>
<li class="chapter" data-level="8.2.8" data-path="map-making-in-R.html"><a href="map-making-in-R.html#true-zoom."><i class="fa fa-check"></i><b>8.2.8</b> True zoom.</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="map-making-in-R.html"><a href="map-making-in-R.html#ggmap-hooray"><i class="fa fa-check"></i><b>8.3</b> ggmap</a><ul>
<li class="chapter" data-level="8.3.1" data-path="map-making-in-R.html"><a href="map-making-in-R.html#three-examples"><i class="fa fa-check"></i><b>8.3.1</b> Three examples</a></li>
<li class="chapter" data-level="8.3.2" data-path="map-making-in-R.html"><a href="map-making-in-R.html#how-ggmap-works"><i class="fa fa-check"></i><b>8.3.2</b> How ggmap works</a></li>
<li class="chapter" data-level="8.3.3" data-path="map-making-in-R.html"><a href="map-making-in-R.html#sisquoctober"><i class="fa fa-check"></i><b>8.3.3</b> Sisquoctober</a></li>
<li class="chapter" data-level="8.3.4" data-path="map-making-in-R.html"><a href="map-making-in-R.html#big-creek-in-big-sur"><i class="fa fa-check"></i><b>8.3.4</b> Big Creek in Big Sur</a></li>
<li class="chapter" data-level="8.3.5" data-path="map-making-in-R.html"><a href="map-making-in-R.html#how-about-a-bike-ride"><i class="fa fa-check"></i><b>8.3.5</b> How about a bike ride?</a></li>
<li class="chapter" data-level="8.3.6" data-path="map-making-in-R.html"><a href="map-making-in-R.html#fish-sampling-locations"><i class="fa fa-check"></i><b>8.3.6</b> Fish sampling locations</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="map-making-in-R.html"><a href="map-making-in-R.html#in-class-review-and-short-assignment"><i class="fa fa-check"></i><b>8.4</b> In-class review and short assignment</a><ul>
<li class="chapter" data-level="8.4.1" data-path="map-making-in-R.html"><a href="map-making-in-R.html#short-assignment-of-plotting-with-maps"><i class="fa fa-check"></i><b>8.4.1</b> Short assignment of plotting with <code>maps</code></a></li>
<li class="chapter" data-level="8.4.2" data-path="map-making-in-R.html"><a href="map-making-in-R.html#short-assignment-of-plotting-with-ggmap"><i class="fa fa-check"></i><b>8.4.2</b> Short assignment of plotting with ggmap()</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="plotting-spatial-data-with-ggplot.html"><a href="plotting-spatial-data-with-ggplot.html"><i class="fa fa-check"></i><b>9</b> Plotting “Spatial” Data with ggplot</a><ul>
<li class="chapter" data-level="9.1" data-path="plotting-spatial-data-with-ggplot.html"><a href="plotting-spatial-data-with-ggplot.html#point-and-vector-data-types"><i class="fa fa-check"></i><b>9.1</b> Point and Vector Data Types</a><ul>
<li class="chapter" data-level="9.1.1" data-path="plotting-spatial-data-with-ggplot.html"><a href="plotting-spatial-data-with-ggplot.html#some-example-vector-data"><i class="fa fa-check"></i><b>9.1.1</b> Some example vector data</a></li>
<li class="chapter" data-level="9.1.2" data-path="plotting-spatial-data-with-ggplot.html"><a href="plotting-spatial-data-with-ggplot.html#reading-in-vector-spatial-data"><i class="fa fa-check"></i><b>9.1.2</b> Reading in Vector Spatial Data</a></li>
<li class="chapter" data-level="9.1.3" data-path="plotting-spatial-data-with-ggplot.html"><a href="plotting-spatial-data-with-ggplot.html#quick-visualization-of-our-vector-data"><i class="fa fa-check"></i><b>9.1.3</b> Quick Visualization of our Vector Data</a></li>
<li class="chapter" data-level="9.1.4" data-path="plotting-spatial-data-with-ggplot.html"><a href="plotting-spatial-data-with-ggplot.html#the-structure-of-spatialdataframes"><i class="fa fa-check"></i><b>9.1.4</b> The Structure of Spatial*DataFrames</a></li>
<li class="chapter" data-level="9.1.5" data-path="plotting-spatial-data-with-ggplot.html"><a href="plotting-spatial-data-with-ggplot.html#subsetting-spatialdataframes"><i class="fa fa-check"></i><b>9.1.5</b> Subsetting Spatial*DataFrames</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="plotting-spatial-data-with-ggplot.html"><a href="plotting-spatial-data-with-ggplot.html#coordinate-reference-systems"><i class="fa fa-check"></i><b>9.2</b> Coordinate Reference Systems</a></li>
<li class="chapter" data-level="9.3" data-path="plotting-spatial-data-with-ggplot.html"><a href="plotting-spatial-data-with-ggplot.html#plotting-things-with-ggspatial"><i class="fa fa-check"></i><b>9.3</b> Plotting things with ggspatial</a><ul>
<li class="chapter" data-level="9.3.1" data-path="plotting-spatial-data-with-ggplot.html"><a href="plotting-spatial-data-with-ggplot.html#want-a-background"><i class="fa fa-check"></i><b>9.3.1</b> Want a background?</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="plotting-spatial-data-with-ggplot.html"><a href="plotting-spatial-data-with-ggplot.html#spatial-raster-data"><i class="fa fa-check"></i><b>9.4</b> Spatial Raster Data</a></li>
<li class="chapter" data-level="9.5" data-path="plotting-spatial-data-with-ggplot.html"><a href="plotting-spatial-data-with-ggplot.html#wrapping-up"><i class="fa fa-check"></i><b>9.5</b> Wrapping Up</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Case Studies in Reproducible Research: a spring seminar at UCSC</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="week5" class="section level1">
<h1><span class="header-section-number">Chapter 6</span> Week Five Meeting</h1>
<div id="some-things-regarding-peoples-repositories" class="section level2">
<h2><span class="header-section-number">6.1</span> Some things regarding people’s repositories</h2>
<div id="data-compression" class="section level3">
<h3><span class="header-section-number">6.1.1</span> Data Compression</h3>
<p>We have been endorsing <code>.csv</code> as a good format for data, and <em>it is</em>, because it is human-readable and easily parsed into tibbles. However, when you have very long tibbles, it is not necessarily the most space-efficient format. Large <code>.csv</code> files can take up much more space on your hard drive <em>than they should</em>.</p>
<p>What do we mean by “<em>than they should?</em>” in that context? This has to do with how much <em>information</em> is in the file, where information is used in the context of <em>information theory</em>. Often tibbles will have columns that have fairly “redundant” information—for example, in a mutlispecies salmon data set, one column might have entries that are either “Chinook”, “coho”, or “steelhead”. It takes a few bytes to store each one of those words, and if they are used in a column that has millions of rows, that can add up to a lot of space on your hard drive. Colloquially, <em>data compression</em> is the art of finding ways of using short “code-names” for things or patterns that occur frequently in a file, and in so doing, reducing the overall file size.</p>
<p>The consequences of data compression can be profound and wonderful. You can reduce the size of a file, sometimes by an order or magnitude or more. There are a few good choices available for compressing your data (making it smaller.). Note that doing so often makes it a little harder to edit your data set; however, if your data set is not going to change, and it is large, then it makes sense to compress it—especially if it is so big you would rather not (or can’t) put it on GitHub.</p>
<div id="gzip" class="section level4">
<h4><span class="header-section-number">6.1.1.1</span> gzip</h4>
<p>If you are working on a Mac, you have the Unix utility <code>gzip</code>. We will illustrate its use on Katie’s big salmon data set, <code>ASL_merged.csv</code>. Let’s see how big that is. We can use the Unix utility <code>du</code> (stands for “disk usage”).</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="co"># give this command on the Terminal in the directory where the file lives:</span>
<span class="kw">du</span> -h ASL_merged.csv </code></pre></div>
<p>The results comes back:</p>
<pre><code>322M    ASL_merged.csv</code></pre>
<p>Whoa! This file is 332 Megabytes. That is quite large!</p>
<p>However, we can compress it like this:</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">gzip</span> ASL_merged.csv </code></pre></div>
<p>When we do that, it compresses the file and renames it to have a <code>.gz</code> extension on it: <code>ASL_merged.csv.gz</code>. We can then see how big that file is:</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">du</span> -h ASL_merged.csv.gz </code></pre></div>
<p>tells us:</p>
<pre><code>11M ASL_merged.csv.gz</code></pre>
<p>Whoa! We went from 332 Megabytes to 11. It is just 3% of its original size (and small enough that you can safely put it on GitHub).</p>
<p>One very nice feature is that gzipped files can be read in directly by the functions of the <code>readr</code> package. So, for example, <code>read_csv()</code> works just fine on the gzipped version of Katie’s massive salmon data set:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this works the same as it would on the ungzipped file</span>
salmon &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">&quot;data/ASL_merged.csv.gz&quot;</span>)</code></pre></div>
</div>
<div id="xz-comression-with-saverds" class="section level4">
<h4><span class="header-section-number">6.1.1.2</span> <code>xz</code> comression with <code>saveRDS()</code></h4>
<p>Another method that can be even more efficient with tibbles is to store them as R objects using the <code>saveRDS()</code> function with the <code>xz</code> compression option. This has the nice advantage that all the data types of the variables (for example, if you had made factors out of some) will be preserved <em>exactly</em> as they are in the tibble when you save it.</p>
<p>Let’s imagine we have read the tibble into the variable <code>salmon</code>, and all the column types were as we wanted. Then, we could save that tibble directly to a compressed file like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">saveRDS</span>(salmon, <span class="dt">file =</span> <span class="st">&quot;ASL_xz.rds&quot;</span>, <span class="dt">compress =</span> <span class="st">&quot;xz&quot;</span>)</code></pre></div>
<p>Note that <code>compress = &quot;xz&quot;</code> option. Let’s see how that did using <code>du</code> on the Unix terminal:</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">du</span> -h ASL_xz.rds</code></pre></div>
<p>tells us:</p>
<pre><code>3.7M    ASL_xz.rds</code></pre>
<p>Holy Smokes! Only 3.7 Megabytes. That is only 1.1% of its original size. Lovely!</p>
<p>In order to read that tibble back into a variable (named <code>my_var</code>, say) in R, you would use <code>readRDS()</code> like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_var &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;ASL_xz.rds&quot;</span>)</code></pre></div>
<p>Voila!</p>
</div>
</div>
</div>
<div id="a-quick-aside-about-missing-data" class="section level2">
<h2><span class="header-section-number">6.2</span> A quick aside about missing data</h2>
<p>Garrett and Hadley note that “missing values are ‘contagious’: almost any operation involving an unknown value will also be unknown.” This is true for the most part, but there is a vexing inconsistency. Observe</p>
<p>This gives us NA as we would hope it would</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="ot">NA</span> ==<span class="st"> </span><span class="dv">0</span> |<span class="st"> </span><span class="ot">NA</span> ==<span class="st"> </span><span class="dv">1</span></code></pre></div>
<pre><code>## [1] NA</code></pre>
<p>However, this one returns FALSE. What gives?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="ot">NA</span> %in%<span class="st"> </span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>)</code></pre></div>
<pre><code>## [1] FALSE</code></pre>
</div>
<div id="brief-highlights-of-the-joins-chapter" class="section level2">
<h2><span class="header-section-number">6.3</span> Brief Highlights of the Joins Chapter</h2>
<p>You will likely end up using joins all the time. As noted in the book, the <code>left_join()</code> is what you will likely use all the time. In this case you have a “focal” data frame with all the rows (“cases” or “observations”) in the <code>x</code> table that you are going to be wanting to add some columns to. Those columns live in the <code>y</code> table (along with the matching keys).</p>
<p>“<em>The left join should be your default join: use it unless you have a strong reason to prefer one of the others.</em>”</p>
<p>You an do the same things with base R’s <code>merge()</code> function, but it is slower and somewhat harder to express your intent with it. (I’ve always really disliked the <code>merge()</code> function…)</p>
<div id="a-few-thoughts-on-keys" class="section level3">
<h3><span class="header-section-number">6.3.1</span> A few thoughts on keys</h3>
<p>It is always a worthwhile exercise to go through and figure out what the <em>primary key</em> is in a tibble you are working with. It might be that the primary key is a compound key: it defines unique observations by a combination of several variables. Sometimes there is no explicit primary key! It is worthwhile to add a <em>surrogate key</em> in that case.</p>
<p>On the flip side of these issues: when you are compiling your own data set, you might want to spend some time making sure that units that might be relevant to an analysis are explicitly identified in a single column. Here is an example: the NOAA Observer program takes tissue samples from bycatch for genetic analysis. There is a primary key <code>tissue_sample</code> for every tissue sample. However, under some circumstances they take multiple tissue samples from the same individual. But they don’t have a column in the data set with an indvividual ID. So, when the send their samples to people who will genotype them, a lot of individuals are unwittingly genotyped twice. Their response: “Well, isn’t it obvious that if a tissue sample is taken from a fish that was caught on the same <code>vessel</code> on the same <code>day</code> and in the same <code>haul</code>, and is of the same <code>species</code> and has the same recorded <code>length</code> is the same individual?” My response: “<strong>NO! It isn’t!</strong>. Don’t make people use a whole lot of columns to identify things that should be identified in a single column!”</p>
</div>
</div>
<div id="an-example-of-using-some-joins" class="section level2">
<h2><span class="header-section-number">6.4</span> An example of using some joins</h2>
<p>Let’s walk through a simple case that should be familiar to those in the group who have worked with genetic data and have had to deal with the problem of attaching meta data to genetic data coming off a genotyping instrument.</p>
<p>Typically those data come out in a form that can be made into a tibble. Let’s read in a toy example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
genos &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">&quot;inputs/toy_geno.csv&quot;</span>)
genos</code></pre></div>
<pre><code>## # A tibble: 8 × 5
##     bird locus1_a locus1_b locus2_a locus2_b
##    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;
## 1 wiwa01        1        2        3        3
## 2 wiwa02        2        2        4        4
## 3 wiwa03        2        2        3        4
## 4 wiwa04        1        1        4        4
## 5 wiwa05        2        2        3        4
## 6 wiwa06        1        1        4        4
## 7 wiwa07        1        2        3        4
## 8 wiwa08        1        1        3        4</code></pre>
<p>There is a single column (<code>bird</code> in this case) that is the primary key that uniquely identifies individuals. Then each locus gets two columns of data (one for each gene copy in a diploid).</p>
<p>That is all well and good. But now, consider this problem: for a particular analysis we are going to do, we need to have the latitude and longitude coordinates where each bird was sampled. Let’s say that we got these samples from friendly collectors who provided a meta data file that gave us the collection location and the name of the collector. Let’s look at that:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">meta &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">&quot;inputs/toy_meta.csv&quot;</span>)
meta</code></pre></div>
<pre><code>## # A tibble: 10 × 3
##    field_id location collector
##       &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;
## 1    wiwa01     swAZ       joe
## 2    wiwa02     nwAZ      mary
## 3    wiwa03     soCA       ted
## 4    wiwa05     soCA       ted
## 5    wiwa06     swAZ       joe
## 6    wiwa08     noCA      erin
## 7    wiwa09     noCA      erin
## 8    wiwa10     noCA      erin
## 9    wiwa11     noCA      erin
## 10   wiwa12     noCA      erin</code></pre>
<p><strong>Notice:</strong> there are some birds in this data set that we don’t have in the <code>genos</code> tibble. Not only that, but if you look closely, it is missing some birds in <code>genos</code>: they are “wiwa04” and “wiwa07”. Also, note that the column that holds the ID of each bird is called <code>field_id</code> not <code>bird</code>.</p>
<p>Finally, the network of bird sample collectors maintains a data base of all their location codes that looks like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">locations &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">&quot;inputs/toy_locations.csv&quot;</span>)
locations</code></pre></div>
<pre><code>## # A tibble: 6 × 3
##   location   lat   long
##      &lt;chr&gt; &lt;dbl&gt;  &lt;dbl&gt;
## 1     swAZ  32.1 -112.8
## 2     nwAZ  36.8 -113.5
## 3     soCA  33.2 -117.1
## 4     noCA  40.6 -123.9
## 5     soOR  42.9 -123.7
## 6     noOR  45.9 -123.1</code></pre>
<p>Aha! So, what we need to do is associate with each <code>bird</code> a <code>location</code>, and then once we have done with that, we need to associate a <code>lat</code> and a <code>long</code> with those locations. This is the perfect job for a join (two of them, actually).</p>
<p>Note that we are focused on our birds, here, so we want to keep them all around and not add any information where we don’t have a bird. Hence <code>left_join()</code> is our go-to friend there (as it almost always will be).</p>
<p>Here is what the first step looks like:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genos %&gt;%
<span class="st">  </span><span class="kw">left_join</span>(., meta, <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;bird&quot;</span> =<span class="st"> &quot;field_id&quot;</span>))</code></pre></div>
<pre><code>## # A tibble: 8 × 7
##     bird locus1_a locus1_b locus2_a locus2_b location collector
##    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;    &lt;chr&gt;     &lt;chr&gt;
## 1 wiwa01        1        2        3        3     swAZ       joe
## 2 wiwa02        2        2        4        4     nwAZ      mary
## 3 wiwa03        2        2        3        4     soCA       ted
## 4 wiwa04        1        1        4        4     &lt;NA&gt;      &lt;NA&gt;
## 5 wiwa05        2        2        3        4     soCA       ted
## 6 wiwa06        1        1        4        4     swAZ       joe
## 7 wiwa07        1        2        3        4     &lt;NA&gt;      &lt;NA&gt;
## 8 wiwa08        1        1        3        4     noCA      erin</code></pre>
<p>Notice that we have some NAs for birds that are not in the meta data. That is the behavior we expect from <code>left_join()</code>: it is not going to discard some of your birds, just because they don’t appear in the meta data.</p>
<p>Note also that when we explicitly give the names of the keys (which differ in the different tibbles) the one on the left corresponds to the <code>x</code> argument to <code>left_join()</code>. Also, notice that these key names <em>must be quoted</em>!! (It is easy to forget that, because you so seldom need quotation marks around things in the tidyverse.)</p>
<p>Now, we can add the lat-longs on there. We will show how that is done by chaining onto the previous command:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genos %&gt;%
<span class="st">  </span><span class="kw">left_join</span>(., meta, <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;bird&quot;</span> =<span class="st"> &quot;field_id&quot;</span>)) %&gt;%
<span class="st">  </span><span class="kw">left_join</span>(., locations)</code></pre></div>
<pre><code>## Joining, by = &quot;location&quot;</code></pre>
<pre><code>## # A tibble: 8 × 9
##     bird locus1_a locus1_b locus2_a locus2_b location collector   lat
##    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;    &lt;chr&gt;     &lt;chr&gt; &lt;dbl&gt;
## 1 wiwa01        1        2        3        3     swAZ       joe  32.1
## 2 wiwa02        2        2        4        4     nwAZ      mary  36.8
## 3 wiwa03        2        2        3        4     soCA       ted  33.2
## 4 wiwa04        1        1        4        4     &lt;NA&gt;      &lt;NA&gt;    NA
## 5 wiwa05        2        2        3        4     soCA       ted  33.2
## 6 wiwa06        1        1        4        4     swAZ       joe  32.1
## 7 wiwa07        1        2        3        4     &lt;NA&gt;      &lt;NA&gt;    NA
## 8 wiwa08        1        1        3        4     noCA      erin  40.6
## # ... with 1 more variables: long &lt;dbl&gt;</code></pre>
<p>Voila! That is what we wanted. Now, we could filter out those NAs and drop the <code>collector</code> column, if desired.</p>
<p>Notice that, since the <code>location</code> column was named <code>location</code> in tibble, <code>left_join()</code> just used that.</p>
<div id="when-would-i-use-right_join" class="section level3">
<h3><span class="header-section-number">6.4.1</span> When would I use <code>right_join()</code>?</h3>
<p>The only time I use this is when I want to add columns to the beginning of a tibble, but I want to preserve all the keys in the table that is going to end up with its columns on the right hand side of the table. And even then I usually just use <code>select</code> after a <code>left_join()</code>. Perhaps an example will be best: for some purposes it is best to keep the genotype data all together on the right hand side of the tibble (often genotype data can take up lots of columns and you might want to be able to see the columns you have joined on without using View() and scrolling way over). In this case you can do this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genos %&gt;%
<span class="st">  </span><span class="kw">right_join</span>(meta, ., <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;field_id&quot;</span> =<span class="st"> &quot;bird&quot;</span>)) %&gt;%
<span class="st">  </span><span class="kw">right_join</span>(locations, .) </code></pre></div>
<pre><code>## Joining, by = &quot;location&quot;</code></pre>
<pre><code>## # A tibble: 8 × 9
##   location   lat   long field_id collector locus1_a locus1_b locus2_a
##      &lt;chr&gt; &lt;dbl&gt;  &lt;dbl&gt;    &lt;chr&gt;     &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;
## 1     swAZ  32.1 -112.8   wiwa01       joe        1        2        3
## 2     nwAZ  36.8 -113.5   wiwa02      mary        2        2        4
## 3     soCA  33.2 -117.1   wiwa03       ted        2        2        3
## 4     &lt;NA&gt;    NA     NA   wiwa04      &lt;NA&gt;        1        1        4
## 5     soCA  33.2 -117.1   wiwa05       ted        2        2        3
## 6     swAZ  32.1 -112.8   wiwa06       joe        1        1        4
## 7     &lt;NA&gt;    NA     NA   wiwa07      &lt;NA&gt;        1        2        3
## 8     noCA  40.6 -123.9   wiwa08      erin        1        1        3
## # ... with 1 more variables: locus2_b &lt;int&gt;</code></pre>
<p>Notice that when you do this, the column name of the <code>x</code> variable, <code>field_id</code> is the one that gets retained.</p>
</div>
<div id="what-about-that-inner_join" class="section level3">
<h3><span class="header-section-number">6.4.2</span> What about that <code>inner_join()</code></h3>
<p>If you knew ahead of time that you couldn’t use any birds that you didn’t have lat-longs for, you could start with an <code>inner_join()</code>, because that would discard birds that don’t have an entry in the meta data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genos %&gt;%
<span class="st">  </span><span class="kw">inner_join</span>(., meta, <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;bird&quot;</span> =<span class="st"> &quot;field_id&quot;</span>))</code></pre></div>
<pre><code>## # A tibble: 6 × 7
##     bird locus1_a locus1_b locus2_a locus2_b location collector
##    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;    &lt;chr&gt;     &lt;chr&gt;
## 1 wiwa01        1        2        3        3     swAZ       joe
## 2 wiwa02        2        2        4        4     nwAZ      mary
## 3 wiwa03        2        2        3        4     soCA       ted
## 4 wiwa05        2        2        3        4     soCA       ted
## 5 wiwa06        1        1        4        4     swAZ       joe
## 6 wiwa08        1        1        3        4     noCA      erin</code></pre>
<p>But, it will probably be easier to follow if you <em>explicitly discard those birds</em> using <code>filter()</code> or by doing a filtering join.</p>
</div>
<div id="an-anti_join-example" class="section level3">
<h3><span class="header-section-number">6.4.3</span> An anti_join example</h3>
<p>If you want to get all the genotype data for birds that don’t occur in the meta data you can use the filtering <code>anti_join()</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genos %&gt;%
<span class="st">  </span><span class="kw">anti_join</span>(., meta, <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;bird&quot;</span> =<span class="st"> &quot;field_id&quot;</span>))</code></pre></div>
<pre><code>## # A tibble: 2 × 5
##     bird locus1_a locus1_b locus2_a locus2_b
##    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;
## 1 wiwa07        1        2        3        4
## 2 wiwa04        1        1        4        4</code></pre>
<p>And to return only those rows for birds that are in the meta data, you could use <code>semi_join()</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genos %&gt;%
<span class="st">  </span><span class="kw">semi_join</span>(., meta, <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;bird&quot;</span> =<span class="st"> &quot;field_id&quot;</span>))</code></pre></div>
<pre><code>## # A tibble: 6 × 5
##     bird locus1_a locus1_b locus2_a locus2_b
##    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;
## 1 wiwa01        1        2        3        3
## 2 wiwa02        2        2        4        4
## 3 wiwa03        2        2        3        4
## 4 wiwa05        2        2        3        4
## 5 wiwa06        1        1        4        4
## 6 wiwa08        1        1        3        4</code></pre>
<p>Note that we could turn it around in order to see which birds are in the meta data, but which don’t occur in the genos:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genos %&gt;%
<span class="st">  </span><span class="kw">anti_join</span>(meta, ., <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;field_id&quot;</span> =<span class="st"> &quot;bird&quot;</span>))</code></pre></div>
<pre><code>## # A tibble: 4 × 3
##   field_id location collector
##      &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;
## 1   wiwa12     noCA      erin
## 2   wiwa11     noCA      erin
## 3   wiwa10     noCA      erin
## 4   wiwa09     noCA      erin</code></pre>
<p>Dammit Erin! You always forget to send us the friggin’ samples! Clearly smokin’ too much of the kind green there in noCA. (Note. These names really are totally fictitious.)</p>
<p>Quick quiz: why would this not work:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genos %&gt;%
<span class="st">   </span><span class="kw">anti_join</span>(meta, ., <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;bird&quot;</span> =<span class="st"> &quot;field_id&quot;</span>))</code></pre></div>
</div>
<div id="just-for-fun-lets-see-a-full_join" class="section level3">
<h3><span class="header-section-number">6.4.4</span> Just for fun, let’s see a full_join()</h3>
<p>This bad boy makes a row with NAs for values in <em>either</em> tibble that are not matched in the other:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genos %&gt;%
<span class="st">  </span><span class="kw">full_join</span>(., meta, <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;bird&quot;</span> =<span class="st"> &quot;field_id&quot;</span>)) %&gt;%
<span class="st">  </span><span class="kw">full_join</span>(., locations) %&gt;%
<span class="st">  </span><span class="kw">print</span>(<span class="dt">n =</span> <span class="dv">20</span>) <span class="co"># so all rows print</span></code></pre></div>
<pre><code>## Joining, by = &quot;location&quot;</code></pre>
<pre><code>## # A tibble: 14 × 9
##      bird locus1_a locus1_b locus2_a locus2_b location collector   lat
##     &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;    &lt;int&gt;    &lt;chr&gt;     &lt;chr&gt; &lt;dbl&gt;
## 1  wiwa01        1        2        3        3     swAZ       joe  32.1
## 2  wiwa02        2        2        4        4     nwAZ      mary  36.8
## 3  wiwa03        2        2        3        4     soCA       ted  33.2
## 4  wiwa04        1        1        4        4     &lt;NA&gt;      &lt;NA&gt;    NA
## 5  wiwa05        2        2        3        4     soCA       ted  33.2
## 6  wiwa06        1        1        4        4     swAZ       joe  32.1
## 7  wiwa07        1        2        3        4     &lt;NA&gt;      &lt;NA&gt;    NA
## 8  wiwa08        1        1        3        4     noCA      erin  40.6
## 9  wiwa09       NA       NA       NA       NA     noCA      erin  40.6
## 10 wiwa10       NA       NA       NA       NA     noCA      erin  40.6
## 11 wiwa11       NA       NA       NA       NA     noCA      erin  40.6
## 12 wiwa12       NA       NA       NA       NA     noCA      erin  40.6
## 13   &lt;NA&gt;       NA       NA       NA       NA     soOR      &lt;NA&gt;  42.9
## 14   &lt;NA&gt;       NA       NA       NA       NA     noOR      &lt;NA&gt;  45.9
## # ... with 1 more variables: long &lt;dbl&gt;</code></pre>
<p>This is not typically what we want! Sometimes it is…but usually you will be using <code>left_join()</code>.</p>
</div>
</div>
<div id="working-with-r-notebooks" class="section level2">
<h2><span class="header-section-number">6.5</span> Working with R Notebooks</h2>
<p>R Notebooks are totally awesome! They combine the nice features of working at the R console (namely having access to variables that remain in your .GlobalEnvironment), with the beauty of being able to document things in an easy to read and digest RMarkdown format.</p>
<p>Here, you can download an example of a short Notebook from one of the projects that Kristen and Eric are working on: <a href="https://www.dropbox.com/s/buq39n64m8524ff/02-choosing-96-SNPs.nb.html?dl=1">choosing-snps</a></p>
<p>R notebooks are RMarkdown documents using the <code>html_document</code> option. They are great for doing and explaining analyses. This is where I end up doing most of my analyses these days. Typically I put them in a directory called <code>R-main</code> in my project. This is a more appropriate place to put long analyses than in README.Rmd. The README.Rmd file should be reserved for describing your data and giving people instructions on how to conduct the analysis, e.g. “Open ./R-main/01-clean-data.Rmd and run it all. Then run all the code in 02-compute-statistics.Rmd, etc.”</p>
<div id="open-your-own-r-notebook" class="section level3">
<h3><span class="header-section-number">6.5.1</span> Open your own R Notebook</h3>
<p>This is easy in RStudio: File -&gt; New File -&gt; R Notebook. This gives you a simple template that lets you see how things work and into which you can insert your own thoughts and writings.</p>
<p>For quick help with formatting: Help -&gt; Markdown Quick Reference</p>
<p>For more Markdown info: Help -&gt; Cheatsheets -&gt; R Markdown Cheat Sheet</p>
</div>
<div id="working-with-r-notebooks-1" class="section level3">
<h3><span class="header-section-number">6.5.2</span> Working with R Notebooks</h3>
<p>Big difference from “regular” R Markdown documents: there is no “Knit” button.</p>
<p>Instead, to get results from the code, you must evaluate it, then “Preview”.</p>
<p>Ways of evaluating code blocks:</p>
<ol style="list-style-type: decimal">
<li>right-facing triangle – evaluate current block</li>
<li>down-facing gray triangle – evaluate all the blocks <em>above</em> this one.</li>
</ol>
<p>Or you can use keyboard shortcuts, or use the “Run” button in the upper right of the document.</p>
<p>Results from code blocks get presented in the Notebook.</p>
<p>To get something like “Knit”: do this:</p>
<p>Run -&gt; Restart R and Run All Chunks</p>
</div>
<div id="some-caveats-about-notebooks" class="section level3">
<h3><span class="header-section-number">6.5.3</span> Some caveats about notebooks</h3>
<ol style="list-style-type: decimal">
<li>You should restart and Run All occasionally to make sure it is reproducible.</li>
<li>When evaluating code within an R notebook, by default the working directory for R is set to the directory that the R notebook live in, not the root directory of the project. So, it can give different results (for example when doing file acess) than the R Console. This can be hugely frustrating.</li>
<li>your variables all live in the GlobalEnvironment, so they are at risk of getting overwritten if you use the same variable name in another Notebook that you are working on at the moment. For this reason, to check reproducibility, occasionaly check that Run -&gt; Restart R and Run All Chunks works for you.</li>
</ol>
<p>The .nb.html files don’t play very well with GitHub. If you want to share them, they are great for emailing to people (but tell them to download it and view it as a file—the gmail viewer does a crappy job of rendering it.)</p>
</div>
<div id="opening-a-.nb.html-file-in-rstudio" class="section level3">
<h3><span class="header-section-number">6.5.4</span> Opening a .nb.html file in Rstudio</h3>
<p>Doing this “reconstitutes” the .Rmd file that made it (along with the results that are saved in it). Which is sort of cool. However, it does not reconstitute all the data, etc. that went into it. So, y’all wouldn’t be able to run 02-choosing-96-SNPs.Rmd.</p>
</div>
</div>
<div id="for-next-week" class="section level2">
<h2><span class="header-section-number">6.6</span> For Next Week:</h2>
<p>Read Chapter 3: Data Visualization</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="week4.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="ggplot2-intro-lecture.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/07-week-5-meeting-joins-etc.Rmd",
"text": "Edit"
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
